### PR TITLE
I161 t2c intents refactor

### PR DESCRIPTION
**Pull request by tomasmacek**
This is major T2C refactor

### CODE REVIEW

File: scripts/DialogData.py [hunk 1]
```diff
@@ -14,44 +14,193 @@
...
-X_PLACEHOLDER = '<x>'
+from EntityData import EntityData
+from NodeData import NodeData
+import unicodedata, unidecode
+from wawCommons import printf, eprintf, toIntentName
+X_PLACEHOLDER = u'&lt;x&gt;'
 
 class DialogData(object):
-    """ Represents complete Dialog data structure. """
+    """ DialogData represents complete dialog (intents, entities, nodes, labels).
```
Comment (Tereza-Pytelova) [inline scripts/DialogData.py]: It is not clear to me what `labels` are.

File: scripts/DialogData.py [hunk 2]
```diff
@@ -14,44 +14,193 @@
...
+    The data are collected across all processed T2C source.
```
Comment (Tereza-Pytelova) [inline scripts/DialogData.py]:
```suggestion
        The data are collected across all processed T2C source.
```

File: scripts/DialogData.py [hunk 3]
```diff
@@ -14,44 +14,193 @@
...
+    #  ENTITY
+    #******************************************
 
-    def getIntentData(self, intentName, domainName=None):
-        if domainName is not None:
-            if domainName not in self._domains:
-                self._domains[domainName] = []
-            if intentName not in self._domains[domainName]:
-                self._domains[domainName].append(intentName)
+    def createEntity(self, entity_name):
```
Comment (Tereza-Pytelova) [inline scripts/DialogData.py]: Why this function behaves differently than the `createIntent` one? (Returns None)

File: scripts/DialogData.py [hunk 4]
```diff
@@ -14,44 +14,193 @@
...
+                    printf('INFO: Label:%s not resolve, expecting tahat label is external node_name\n', label)
+
+#   DOMAINS
+#******************************************
+    def getAllDomains(self):
         return self._domains
+
+# createUnique ..
+#******************************************
+    def createUniqueIntentName(self, intent_name):
```
Comment (Tereza-Pytelova) [inline scripts/DialogData.py]: This should be merged with `toIntentName` function: https://github.com/IBM/watson-assistant-workbench/blob/7d164be90adae79aebd6fdd0740d9885ad57c3cb/scripts/wawCommons.py#L62

File: scripts/DialogData.py [hunk 5]
```diff
@@ -14,44 +14,193 @@
...
+            return unique_intent_name
+        #try to modify by a number
+        for modifier in range(0, 10000):
+            new_unique_intent_name=unique_intent_name+repr(modifier)  #create a modified one
+            # Check if the name exists
+            if new_unique_intent_name not in self._intents:
+                return new_unique_intent_name
+        return None
+
+    def createUniqueEntityName(self, entity_name):
```
Comment (Tereza-Pytelova) [inline scripts/DialogData.py]: As the previous function - please merge it with `toEntityName` function:  https://github.com/IBM/watson-assistant-workbench/blob/7d164be90adae79aebd6fdd0740d9885ad57c3cb/scripts/wawCommons.py#L131

File: scripts/DialogData.py (hunk 3 above)
Comment (tomasmacek) [reply scripts/DialogData.py]: Agree, code was unified and checks added for situations where attempts to create nodes, intents or entities of the same name are made

File: scripts/DialogData.py (hunk 4 above)
Comment (tomasmacek) [reply scripts/DialogData.py]: createUniqueIntentName is based on toIntentName but adds the checks for existence of the name and auto generation of the modifies to make it unique. Don't we need both of them? Let's discuss further.

File: scripts/XLSXHandler.py [hunk 6]
```diff
@@ -232,6 +232,9 @@ def __handle_x_condition_block(self, block, domain, label):
             node_condition = re.sub(Dialog.X_PLACEHOLDER, row[0], block_header)
             node_name = self._dialogData.createUniqueNodeName(node_condition)  # make it unique
             nodeData = self._dialogData.createNode(node_name, domain)  # create space for new node, remembers node_name
+            if nodeData is None:
```
Comment (Tereza-Pytelova) [inline scripts/XLSXHandler.py]: I would move this error into the `createNode` function.

File: scripts/DialogData.py (hunk 5 above)
Comment (tomasmacek) [reply scripts/DialogData.py]: As discussed, the part of stripping accents was based on toIntentName, the rest left for further refactoring

File: scripts/EntityData.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+import re
+from wawCommons import eprintf
+from collections import OrderedDict
+
+class EntityData(object):
+    """ Represents a single entity.    """
```
[the remark this answers is not in the document]
Comment (tomasmacek) [reply scripts/EntityData.py]: Synonyms are part of value in T2C, it is a list of items separated by ; the first one is the value

File: scripts/NodeData.py [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+from wawCommons import eprintf
+from collections import OrderedDict
+
+class NodeData(object):
+    """ Represents a data structure containing all necessary information for a single dialog node
+    """
+
+    def __init__(self):
+        self._channels = {}             # key: channel name, value: list of all outputs for the channel
+        self._variables = {}            # key: variable name, value: variable value
```
[the remark this answers is not in the document]
Comment (tomasmacek) [reply scripts/NodeData.py]: Adding more comments which are still brief as expecting that programmer will check user manual of T2C first.

File: scripts/NodeData.py [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+                else:
+                    self.__handleChannelDefinition(item)
+
+        if len(rawOutputs) >= 2:
+            if rawOutputs[1]:
+                self.__handleButtonDefinition(rawOutputs[1])
+            if rawOutputs[2]:
+                self.__handleJumpToDefinition(rawOutputs[2], labelsMap)
+
+    def generateNodes(self):
```
[the remark this answers is not in the document]
Comment (tomasmacek) [reply scripts/NodeData.py]: good spot! removed

File: scripts/wawCommons.py [hunk 10]
```diff
@@ -112,16 +113,22 @@ def toIntentName(NAME_POLICY, userReplacements, *intentSubnames):
                 uIntentSubnameUser = uNewIntentSubnameUser
             if uIntentSubnameUser != uIntentSubnameWA: # user restriction triggered
                 restrictionTextIntentName.append("User-defined regex: '" + "', '".join(triggeredUserRegex) + "'.")
-        if uIntentSubnameUser != uIntentSubname:
```
[the remark this answers is not in the document]
Comment (tomasmacek) [reply scripts/wawCommons.py]: The problem was that it flooded log with warning messages. I am returning it back but commenting out wardings for the time being till we  get better logging in place,

File: scripts/XLSXHandler.py (hunk 6 above)
Comment (tomasmacek) [reply scripts/XLSXHandler.py]: this is now the same for createNode, createEntity and createIntent - it wold requre change in all the places
Are there any arguments to make the change? The general push is usually to log in higher levels  as there might be reason in the future to mask the error (though I am not sure at the moment why). Let's discuss.

File: scripts/EntityData.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Tereza-Pytelova) [reply scripts/EntityData.py]: I would write this information somewhere into code.

[the remark this answers is not in the document]
Comment (tomasmacek) [reply scripts/EntityData.py]: done

File: scripts/XLSXHandler.py (hunk 6 above)
Comment (Tereza-Pytelova) [reply scripts/XLSXHandler.py]: After discussion: move just info print, leave `exit()` here.

File: scripts/wawCommons.py [hunk 11]
```diff
@@ -79,6 +79,7 @@ def toIntentName(NAME_POLICY, userReplacements, *intentSubnames):
         uIntentSubname = intentSubname.decode('utf-8') if isinstance(intentSubname, str) else intentSubname
         # apply WA restrictions (https://console.bluemix.net/docs/services/conversation/intents.html#defining-intents)
         uIntentSubnameWA = re.sub(' ', '_', uIntentSubname, re.UNICODE) # replace space by underscore
```
[the remark this answers is not in the document]
Comment (Tereza-Pytelova) [reply scripts/wawCommons.py]: And please remove following line ;)

**Pull request merged**